File: 2020/day03/day03.py
import sys

fname = '../input/day03/input.txt'

part1 = True
if len(sys.argv) > 1 and sys.argv[1] == '2':
    part1 = False

size_x = 0
size_y = 0
mmap = []
lines = []

with open(fname) as fh:
    lines = fh.readlines()

row = 0
for line in lines:
    line = line.strip()
    size_x = len(line)
    col = 0
    mmap.append([])
    for x in line:
        mmap[row].append(x)
        col = col + 1
    row = row + 1
size_y = row

def slope(step_x, step_y):
    x = 0
    y = 0
    num = 0
    while y < size_y - 1:
        x = x + step_x
        y = y + step_y
        if x >= size_x:
            x = x - size_x
        if mmap[y][x] == '#':
            num = num + 1
    return num

if part1:
    num = slope(3, 1)
    print(num)
    sys.exit(0)

n1 = slope(1, 1)
n2 = slope(3, 1)
n3 = slope(5, 1)
n4 = slope(7, 1)
n5 = slope(1, 2)

print(n1, n2, n3, n4, n5)
print(n1*n2*n3*n4*n5)

File: 2020/java/aoc2020/src/main/java/org/f5n/aoc2020/days/Day03.java
package org.f5n.aoc2020.days;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.f5n.aoc2020.utils.*;


public class Day03 extends Day {
	protected List<String> input;
	protected char[][] mmap = new char[1][1];
	protected int sx = 0;
	protected int sy = 0;

	public Day03(List<String> input) {
		this.input = input;
	}

	protected Result part1(Input limit) {
		prep();
		long rv = slope(3, 1);
		return new IntResult(rv);
	}

	protected Result part2(Input limit) {
		prep();
		long rv = slope(1, 1) * slope(3, 1) * slope(5, 1) * slope(7, 1) * slope(1, 2);
		return new IntResult(rv);
	}

	private void prep() {
		sy = input.size();
		sx = input.get(0).length();
		mmap = new char[sy+1][sx+1];
		int i = 0;
		for (String line : input) {
			line = line.strip();
			for (int j = 0; j < sx; ++j) {
				mmap[i][j] = line.charAt(j);
			}
			++i;
		}
	}

	private long slope(int stepx, int stepy) {
		int x = 0;
		int y = 0;
		int num = 0;
		while (y < sy - 1) {
			x += stepx;
			y += stepy;
			if (x >= sx) {
				x -= sx;
			}
			if (mmap[y][x] == '#') {
				++num;
			}
		}
		return num;
	}
}